controller/main.go

```diff
@@ -71,38 +71,26 @@ func (c *controller) SetBalancer(l log.Logger, name string, svcRo *v1.Service, _
 	svc := svcRo.DeepCopy()
 	successRes := controllers.SyncStateSuccess
 	wasAllocated := c.isServiceAllocated(name)
-	if !c.convergeBalancer(l, name, svc) {
-		return controllers.SyncStateError
-	}
+	c.convergeBalancer(l, name, svc)
+
 	if wasAllocated && !c.isServiceAllocated(name) { // convergeBalancer may deallocate our service and this means it did it.
 		// if the service was deallocated, it may have have left room
 		// for another one, so we reprocess
 		level.Info(l).Log("event", "serviceUpdated", "msg", "removed loadbalancer from service, services will be reprocessed")
 		successRes = controllers.SyncStateReprocessAll
 	}
-	if reflect.DeepEqual(svcRo, svc) {
-		level.Debug(l).Log("event", "noChange", "msg", "service converged, no change")
-		return successRes
-	}
-
-	toWrite := svcRo.DeepCopy()
-	if !reflect.DeepEqual(svcRo.Status, svc.Status) {
-		toWrite.Status = svc.Status
-	}
-
-	if !reflect.DeepEqual(svcRo.Annotations, svc.Annotations) {
-		toWrite.Annotations = svc.Annotations
-	}
 
-	if !reflect.DeepEqual(toWrite, svcRo) {
+	// Only the Status and Annotations are modified in convergeBalancer.
+	if !reflect.DeepEqual(svcRo.Status, svc.Status) || !reflect.DeepEqual(svcRo.Annotations, svc.Annotations) {
 		if err := c.client.UpdateStatus(svc); err != nil {
 			level.Error(l).Log("op", "updateServiceStatus", "error", err, "msg", "failed to update service")
 			return controllers.SyncStateError
 		}
 		level.Info(l).Log("event", "serviceUpdated", "msg", "updated service object")
+	} else {
```

Comment:
I'd rather have an early return than an else here

Reply:
Agree, it looks better. However, I didn't put it because it's so close to the end that it is not an early return anymore. Will change it.